tests/dynrpg.cpp

```diff
@@ -0,0 +1,204 @@
+#include <lcf/data.h>
+#include "doctest.h"
+#include "dynrpg.h"
+#include "game_variables.h"
+#include "test_mock_actor.h"
+
+TEST_SUITE_BEGIN("DynRPG");
+
+class LogGuard {
+public:
+	LogGuard() {
+		lvl = Output::GetLogLevel();
+		Output::SetLogLevel(LogLevel::Error);
+	}
+
+	~LogGuard() {
+		Output::SetLogLevel(lvl);
+	}
+
+	LogLevel lvl;
+};
+
+TEST_CASE("Basic parsing") {
+	std::vector<std::string> args;
+	CHECK(DynRpg::ParseCommand(R"(@FunC abc , "Hello World", 42, 1.5, 1.5a, "Str"",ing")", args) == "func");
+	CHECK(args.size() == 6);
+	CHECK(args[0] == "abc");
+	CHECK(args[1] == "Hello World");
+	CHECK(args[2] == "42");
+	CHECK(args[3] == "1.5");
+	CHECK(args[4] == "1.5a"); // DynRPG returns 1.5, not critical, we extract float later
+	CHECK(args[5] == "Str\",ing");
+}
+
+TEST_CASE("Unterminated literal") {
+	std::vector<std::string> args;
+	CHECK(DynRpg::ParseCommand("@FunC abc , \"Hello ", args) == "func");
+	CHECK(args.size() == 2);
+	CHECK(args[0] == "abc");
+	CHECK(args[1] == "Hello ");
+}
+
+TEST_CASE("Comma") {
+	std::vector<std::string> args;
+	CHECK(DynRpg::ParseCommand("@FunC , abc,,,xyz,,", args) == "func");
+	CHECK(args.size() == 7);
+	CHECK(args[0] == "");
+	CHECK(args[1] == "abc");
+	CHECK(args[2] == "");
+	CHECK(args[3] == "");
+	CHECK(args[4] == "xyz");
+	CHECK(args[5] == "");
+	CHECK(args[6] == "");
+}
+
+TEST_CASE("Tokens") {
+	std::vector<std::string> args;
+	CHECK(DynRpg::ParseCommand("@FunC ab cd,Bla  ,   Bla Blub, Ab\"C \"d e", args) == "func");
+	CHECK(args.size() == 4);
+	CHECK(args[0] == "abcd");
+	CHECK(args[1] == "bla");
+	CHECK(args[2] == "blablub");
+	CHECK(args[3] == "ab\"c\"de");
+}
+
+TEST_CASE("Variable names") {
+	Main_Data::game_variables = std::make_unique<Game_Variables>(-99999, 99999);
+	std::vector<int32_t> vars = {100, 4, 2, 1};
+	Main_Data::game_variables->SetData(vars);
+	Main_Data::game_variables->SetWarning(0);
+
+	std::vector<std::string> args;
+	CHECK(DynRpg::ParseCommand("@FunC V2,VV3,VVVV3", args) == "func");
+	CHECK(args.size() == 3);
+	CHECK(args[0] == "4");
+	CHECK(args[1] == "4");
+	CHECK(args[2] == "100");
+
+	Output::SetLogLevel(LogLevel::Error);
+}
+
+TEST_CASE("Variable names with junk") {
+	Main_Data::game_variables = std::make_unique<Game_Variables>(-99999, 99999);
```

Comment:
Lets allocate `game_variables` in `MockActor` and use it here. That way the variables get destroyed when this test exits.

Since all unit tests run under one binary, we have to be super careful about resetting all global state.

Once I've refactored singletons for Player, I'll also refactor for tests. For now we're stuck with this messy half situation.